src/common/strtol.cc

```diff
@@ -128,69 +147,185 @@ float strict_strtof(const char *str, std::string *err)
   return ret;
 }
 
+float strict_strtof(const char *str, std::string *err)
+{
+  return strict_strtof(std::string_view(str), err);
+}
+
 template<typename T>
-T strict_si_cast(const char *str, std::string *err)
+T strict_iec_cast(const std::string_view str, std::string *err)
 {
-  std::string s(str);
-  if (s.empty()) {
-    *err = "strict_sistrtoll: value not specified";
+  if (str.empty()) {
+    *err = "strict_iecstrtoll: value not specified";
     return 0;
   }
-  const char &u = s.back();
+  // get a view of the unit and of the value
+  std::string_view unit;
+  std::string_view n = str;
+  size_t u = str.find_first_not_of("0123456789-+");
   int m = 0;
-  if (u == 'B')
-    m = 0;
-  else if (u == 'K')
-    m = 10;
-  else if (u == 'M')
-    m = 20;
-  else if (u == 'G')
-    m = 30;
-  else if (u == 'T')
-    m = 40;
-  else if (u == 'P')
-    m = 50;
-  else if (u == 'E')
-    m = 60;
-  else
-    m = -1;
-
-  if (m >= 0)
-    s.pop_back();
-  else
-    m = 0;
-
-  long long ll = strict_strtoll(s.c_str(), 10, err);
+  // deal with unit prefix is there is one
+  if (u != std::string_view::npos) {
+    n = str.substr(0, u);
+    unit = str.substr(u, str.length() - u);
+    // we accept both old si prefixes as well as the proper iec prefixes
+    // i.e. K, M, ... and Ki, Mi, ...
+    if (unit.back() == 'i') {
+      if (unit.front() == 'B') {
+        *err = "strict_iecstrtoll: illegal prefix \"Bi\"";
+        return 0;
+      }
+    }
+    if (unit.length() > 2) {
+      *err = "strict_iecstrtoll: illegal prefix (length > 2)";
+      return 0;
+    }
+    if (unit.front() == 'K')
```

Comment:
It's fine as-written, but this might be more tidy written as a switch() statement.

Reply:
@jan--f i think a switch case would be better.